pkgs/stdenv/darwin/override-sdk.nix

```diff
@@ -0,0 +1,412 @@
+          ))
+        ];
+
+        passAsFile = [ "dependencies" ];
+      };
+    in
+    # Only remap the package’s propagated inputs if there are any and if any of them
+    # had packages remapped (with frameworks or proxy packages).
+    if
+      lib.isDerivation pkg && lib.any (inputs: inputs != [ ]) propagatedInputs && env.dependencies != ""
+    then
+      mkStub pkg.name env
+    else
+      pkg;
+
+  propagatedInputs = [
+    "depsBuildBuildPropagated"
+    "propagatedNativeBuildInputs"
+    "depsBuildTargetPropagated"
+    "depsHostHostPropagated"
+    "propagatedBuildInputs"
+    "depsTargetTargetPropagated"
+  ];
+
+  # Gets all propagated inputs in a package. This does not recurse.
+  getPropagatedInputs =
+    pkg:
+    lib.concatMap (input: pkg.${input} or [ ]) [
+      "depsBuildBuildPropagated"
+      "propagatedNativeBuildInputs"
+      "depsBuildTargetPropagated"
+      "depsHostHostPropagated"
+      "propagatedBuildInputs"
+      "depsTargetTargetPropagated"
+    ];
+
+  # Gets all propagated dependencies in a package.
+  getPropagatedDependencies =
+    pkgs:
+    lib.genericClosure {
+      startSet = map (pkg: {
+        key = pkg.outPath;
+        deps = getPropagatedInputs pkg;
+      }) pkgs;
+      operator =
+        { key, deps }:
+        map (dep: {
+          key = dep.outPath;
+          deps = getPropagatedInputs dep;
+        }) deps;
+    };
+
+  # Returns a package mapping based on remapping all propagated packages.
+  #
+  # Implementation note: recursion is avoided because it is costly. Instead, a flat
+  # list of dependencies is obtained then sorted topologically, so that dependencies
+  # can be updated in order (and only once). This produces a new mapping that must
+  # be used to perform the actual updates to the propagated packages.
+  getPackageMapping =
+    baseMapping: input:
+    let
+      dependencies = lib.pipe input [
+        getPropagatedDependencies
+        (lib.toposort (x: y: lib.any (z: x.key == z) y.deps))
+        (lib.getAttr "result")
+        lib.reverseList
+      ];
+    in
+    lib.foldl' (
+      newPackages: pkg:
+      let
+        replacement = replacePropagatedPackages newPackages pkg;
+        outPath = builtins.unsafeDiscardStringContext pkg;
+      in
+      if pkg == null || newPackages ? ${outPath} then
+        newPackages
+      else
+        newPackages // { ${outPath} = replacement; }
+    ) baseMapping input;
+
+  overrideSDK =
+    stdenv: sdkVersion:
+    let
+      newVersion = {
+        inherit (stdenv.hostPlatform) darwinMinVersion darwinSdkVersion;
+      } // (if lib.isAttrs sdkVersion then sdkVersion else { darwinSdkVersion = sdkVersion; });
+
+      inherit (newVersion) darwinMinVersion darwinSdkVersion;
+
+      # Used to get an SDK version corresponding to the requested `darwinSdkVersion`.
+      # TODO: Treat `darwinSdkVersion` as a constraint rather than as an exact version.
```

Comment:
Treating it as a constraint is needed to make updating SDKs not be a breaking change. Otherwise, updating from the 14.0 to the 14.3 SDK would break everything that used `overrideSDK stdenv "14.0";`. The constraint for a bare number would be similar to a `^` constraint with some special handling for Darwin SDK versioning.

For example:`14.0` would match `14.3` but not `15`; `10.13` would match `10.13.6` but not `10.14`.

(Supporting other types of constraints later is possible, but I don’t see a use case for SDK versioning.)

Reply:
The interpretation of a string argument can change, that seems sensible. But I could see a use for `darwinMaxSdkVersion` for example, probably with a name that doesn't imply it means something similar to `darwinMinSdkVersion`.

It's true that the APIs are probably stable enough that it's not very important though.

Reply:
That could be done as a new constraint (such as `<=14` for the newest SDK prior to the 14.0 SDK). I just don’t want to support it without an actual need for it yet.